tooling/templatize/pkg/pipeline/run.go

```diff
@@ -78,56 +82,320 @@ func (o ShellOutput) GetValue(_ string) (*OutPutValue, error) {
 type Outputs map[string]map[string]Output
 
-func RunPipeline(pipeline *types.Pipeline, ctx context.Context, options *PipelineRunOptions) (Outputs, error) {
-	outPuts := make(Outputs)
+type Executor func(s types.Step, ctx context.Context, executionTarget ExecutionTarget, options *PipelineRunOptions, outPuts Outputs) (Output, error)
+
+func RunPipeline(pipeline *types.Pipeline, ctx context.Context, options *PipelineRunOptions, executor Executor) (Outputs, error) {
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	logger.Info("Generating execution graph.")
+	nodes, err := createExecutionGraph(ctx, pipeline, options.Region, options.SubsciptionLookupFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	lock := &sync.RWMutex{}
+	executed := sets.Set[types.StepDependency]{}
+	queued := sets.Set[types.StepDependency]{}
+	outputs := make(Outputs)
+
+	queue := make(chan *stepExecutionNode, len(nodes))
+	done := make(chan struct{}, len(nodes))
+	errs := make(chan error, len(nodes))
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	// producer routine checks to see if we can queue more steps when we finish executing one
+	go func() {
+		thisLogger := logger.WithValues("routine", "producer")
+		defer func() {
+			wg.Done()
+			thisLogger.Info("Producer thread shutting down.")
+		}()
+		for {
+			select {
+			case _, open := <-done:
+				if !open {
+					thisLogger.Info("Done channel closed.")
+					close(queue)
+					return
+				}
+				thisLogger.Info("Processing queue after step finished executing.")
+				lock.RLock()
+				for _, node := range nodes {
+					thisNode := types.StepDependency{
+						ResourceGroup: node.Step.ResourceGroup.Name,
+						Step:          node.Step.Delegate.StepName(),
+					}
+					if queued.Has(thisNode) {
+						continue
+					}
+					if executed.HasAll(contextsToDependencies(node.Parents)...) {
+						thisLogger.Info("Queueing step to run.", "resourceGroup", node.Step.ResourceGroup.Name, "step", node.Step.Delegate.StepName())
+						queued.Insert(thisNode)
+						queue <- node
+					}
+				}
+				lock.RUnlock()
+				thisLogger.Info("Execution status.", "nodes", len(nodes), "queued", len(queued), "executed", len(executed))
+				if len(queued) == len(nodes) {
+					thisLogger.Info("Queued all nodes.")
+					close(queue)
+					return
+				}
+			case <-ctx.Done():
+				thisLogger.Info("Context cancelled.")
+				close(queue)
+				return
+			}
+		}
+	}()
+	// consumer routines pop steps off the execution queue and signal that they finished
+	const maxConcurrency = 1 // TODO: actually do parallel execution and see what breaks
+	for i := 0; i < maxConcurrency; i++ {
+		wg.Add(1)
+		go func() {
+			thisLogger := logger.WithValues("routine", fmt.Sprintf("consumer-%d", i))
+			defer func() {
+				wg.Done()
+				thisLogger.Info("Consumer thread shutting down.")
+			}()
+			for {
+				select {
+				case step, open := <-queue:
+					if !open {
+						thisLogger.Info("Queue channel closed.")
```

Comment:
With more than one consumer routine, this (and all of the early exits closing `done` and `errs`) will cause writes to closed channel panics as the consumers race to close everything.